src/Mesh/Interpolation.jl

```diff
@@ -8,207 +8,460 @@ using CLIMA.Mesh.Geometry
+ - `xbnd` Domain boundaries in x1, x2 and x3 directions
  - `xres` Resolution of the interpolation grid in x1, x2 and x3 directions
 """
-struct InterpolationBrick{FT <:AbstractFloat}
-    realelems::UnitRange{Int64}
-    poly_order::Int
+struct InterpolationBrick{FT  <:AbstractFloat, 
+                           T  <:Int, 
+                         UI8  <:UInt8, 
+                        UI16  <:UInt16, 
+                         I32  <:Int32,
+                         FTV  <:AbstractVector{FT}, 
+                         FTVD <:AbstractVector{FT}, 
+                          TVD <:AbstractVector{T}, 
+                        FTA2  <:Array{FT,2},
+                        UI8AD <:AbstractArray{UI8,2}, 
+                        UI16V <: AbstractVector{UI16},
+                       UI16VD <:AbstractVector{UI16},
+                         I32V <: AbstractVector{I32}}
+    Nel::T         # # of elements
+    Np::T
+    Npl::T
+    poly_order::T
+
+    xbnd::FTA2     # domain bounds, [2(min/max),ndim]
+    xres::FTV      # resolutions in x1, x2, x3 directions for the uniform grid
+
+    x1g::FTV # 1D interpolation grid in x1 direction
+    x2g::FTV
+    x3g::FTV
 
-    xbnd::Array{FT,2}      # domain bounds, [2(min/max),ndim]
-    xres::Array{FT,1}      # resolutions in x1, x2, x3 directions for the uniform grid
+    ξ1::FTVD # unique ξ1 coordinates of interpolation points within each element 
+    ξ2::FTVD # unique ξ2 coordinates of interpolation points within each element 
+    ξ3::FTVD # unique ξ3 coordinates of interpolation points within each element 
 
-    ξ1::Vector{Vector{FT}} # unique ξ1 coordinates of interpolation points within each element 
-    ξ2::Vector{Vector{FT}} # unique ξ2 coordinates of interpolation points within each element 
-    ξ3::Vector{Vector{FT}} # unique ξ3 coordinates of interpolation points within each element 
-  
-    x::Vector{Array{FT,2}} # x[elno][3,npts] -> (x1,x2,x3) coordinates of portion of interpolation grid embedded within each element 
+    flg::UI8AD # flags when ξ1/ξ2/ξ3 interpolation point matches with a GLL point
 
-    V::Vector{Vector{FT}}  # interpolated variable within each element
+    fac::FTVD # normalization factor
+
+    x1i::UI16VD # interpolation grid indices of interpolation points within each element
+    x2i::UI16VD
+    x3i::UI16VD
+
+    offset::TVD  # offsets for each element for v
+
+    m1_r::FTVD  # GLL points
+    m1_w::FTVD  # GLL weights
+    wb::FTVD    # Barycentric weights
+    # MPI setup for gathering interpolated variable on proc # 0
+    Nel_all::I32V 
+ 
+    x1i_all::UI16V
+    x2i_all::UI16V
+    x3i_all::UI16V
 #--------------------------------------------------------
-    function InterpolationBrick(grid::DiscontinuousSpectralElementGrid{FT}, xres) where FT <: AbstractFloat
-        T = Int
+    function InterpolationBrick(grid::DiscontinuousSpectralElementGrid{FT}, xbnd::Array{FT,2}, xres) where FT <: AbstractFloat
+        mpicomm = MPI.COMM_WORLD
```

Comment:
Get this from `grid`
```suggestion
        mpicomm = grid.topology.mpicomm
```